core/metamodel/src/main/java/org/apache/isis/core/metamodel/services/ServicesInjector.java

```diff
@@ -145,16 +149,19 @@ private ServicesInjector(
     }
 
     public boolean isRegisteredService(final Class<?> cls) {
-        // lazily construct cache
-        if(serviceByConcreteType.isEmpty()) {
-            for (Object service : services) {
-                final Class<?> concreteType = service.getClass();
-                serviceByConcreteType.put(concreteType, service);
-            }
-        }
-        return serviceByConcreteType.containsKey(cls);
+        return serviceByConcreteType.get().containsKey(cls);
     }
 
+    public boolean isRegisteredServiceInstance(final Object pojo) {
+        if(pojo==null) {
+            return false;
+        }
+        final Class<?> key = pojo.getClass();
```

Comment:
One thing to be aware of ... when we wrap the domain object then the class will be the synthetic class created by ByteBuddy/Javassist, which will be a subclass of the actual class.  This could be an issue.  There is the concept of a ClassSubstitutor, which was introduced to deal with this: if the class is synthetic, then get its superclass.  I think this mechanism (or something equivalent) should be used everywhere.